libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.h

```diff
@@ -0,0 +1,201 @@
+        return is_healthy_by_index(get_device_index(serialnumber));
+    }
+
+private:
+
+    // Message groups form part of the CAN ID of each frame
+    enum class PacketType : uint8_t {
+        UNKNOWN         = 0,
+        ACK             = 0x21,
+        NACK            = 0x22,
+        PING            = 0x23,
+        FAULT           = 0x42,
+        ALGORITHM_SET   = 0x44,
+        ALGORITHM_GET   = 0x45,
+        STREAM_FAULT    = 0x47,
+        STREAM_INPUT    = 0x48,
+        STREAM_OUTPUT   = 0x49,
+        VOLTAGE_SET     = 0x4B,
+        VOLTAGE_GET     = 0x4C,
+        CVT_SET         = 0x4D,
+        CVT_GET         = 0x4E,
+    };
+
+    enum class Fault_Flags : uint8_t {
+        NONE                = 0,
+        OVER_VOLTAGE        = (1<<0),
+        UNDER_VOLTAGE       = (1<<1),
+        OVER_CURRENT        = (1<<2),
+        OVER_TEMPERATURE    = (1<<3),
+    };
+
+    enum class Config_Mode : int8_t {
+        DISABLED            = 0,
+        REPORT_INPUTS       = 1,
+        REPORT_OUTPUTS      = 2,
+        REPORT_VOUT_CVT_ALG = 3,
+        SET_VOUT            = 4,
+        SET_CVT             = 5,
+        SET_ALGORITHM       = 6,
+    };
+
+    // loop to send output to ESCs in background thread
+    void loop();
+
+    // write frame on CAN bus, returns true on success
+    bool write_frame(AP_HAL::CANFrame &out_frame, const uint64_t timeout);
+
+    // read frame on CAN bus, returns true on success
+    bool read_frame(AP_HAL::CANFrame &recv_frame, const uint64_t timeout, AP_HAL::CANIface::CanIOFlags &flags);
+
+    // parse inbound frames
+    void handle_frame(const AP_HAL::CANFrame &frame);
+
+    // find first MPPT_devices[] index that contains this serialnumber. If not found, returns UINT8_MAX
+    uint8_t get_device_index(const uint16_t serial_number) const;
+
+    // send command to MPPT device
+    void send_command(const PacketType type, const uint16_t serialnumber, const float data = 0.0f);
+
+    bool is_healthy_by_index(const uint8_t index) const {
+        if (index >= device_count) {
+            return false;
+        }
+//        if (MPPT_devices[index].faults != Fault_Flags::NONE) {
+//            // TODO: test this
+//            return false;
+//        }
+        return MPPT_devices[index].is_healthy();
+    }
+
+    void perform_config_tasks();
+    void perform_logging() const;
+
+    const char* get_fault_code_string(const Fault_Flags fault) const;
+
+    //Frames received from the MPPT will use CAN Extended ID: 0x0042XXXX,
+    // the least significant 16 bits contain the serial number of the MPPT.
+    static constexpr uint32_t EXPECTED_FRAME_ID = (0x00420000 | AP_HAL::CANFrame::FlagEFF);
+
+    bool _initialized;
+    char _thread_name[sizeof("PDCAN") + 5];
+
+    AP_HAL::CANIface* _can_iface;
+    HAL_EventHandle _event_handle;
+    uint8_t _driver_index;
+
+
+    HAL_Semaphore _sem;
+
+    uint8_t sequence_prev;
+    static const uint8_t device_count_max = 10;
```

Comment:
So that's a *vast* amount of RAM you're burning there....

Probably better to dynamically allocate.

Reply:
<sigh>Really?
It's 29 bytes each. Putting in the logic for dynamic allocation will burn way more program memory.
How about I reduce it to 5?

Reply:
changed to a define default of 5. That way someone like me can override it without a codechange if necessary.